File: Problems/Algorithms/2131. Longest Palindrome by Concatenating Two Letter Words/LongestPalindromeTwoLetterWords.java
class Solution {
    public int longestPalindrome(String[] words) {
        HashMap<String, Integer> map = new HashMap<>();
        
        int ans = 0;
        for (String word : words) {
            StringBuilder sb = new StringBuilder();
            sb.append(word.charAt(1));
            sb.append(word.charAt(0));
            String inv = sb.toString();
            
            if (map.containsKey(inv) && map.get(inv) > 0) {
                map.put(inv, map.get(inv) - 1);
                ans += 4;
            } else {
                map.put(word, map.getOrDefault(word, 0) + 1);
            }
        }
        
        boolean flag = false;
        for (String word : map.keySet()) {
            if (word.charAt(0) == word.charAt(1) && map.get(word) > 0) {
                flag = true;
            }
        }
        
        if (flag) {
            ans += 2;
        }
        
        return ans;
    }
}

File: Problems/Algorithms/2131. Longest Palindrome by Concatenating Two Letter Words/longest_palindrome_two_letter_words.py
#!/usr/bin/env python3
# coding: utf-8

from collections import defaultdict
from typing import List


class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        words_dict = defaultdict(int)

        ans = 0
        for word in words:
            if word[::-1] in words_dict and words_dict[word[::-1]] > 0:
                words_dict[word[::-1]] -= 1
                ans += 4
            else:
                words_dict[word] += 1

        for word in words_dict:
            if word[0] == word[1] and words_dict[word] > 0:
                ans += 2
                break
        return ans
